src/node_usb.h

```diff
@@ -60,12 +68,23 @@ struct Device: public Napi::ObjectWrap<Device> {
 
 	Napi::Value ClearHalt(const Napi::CallbackInfo& info);
 protected:
-	static std::map<libusb_device*, Device*> byPtr;
-	static Napi::FunctionReference constructor;
-	
 	Napi::Value Constructor(const Napi::CallbackInfo& info);
 };
 
+struct ModuleData {
+	libusb_context* usb_context;
+	std::thread usb_thread;
+
+	bool hotplugEnabled = 0;
+	libusb_hotplug_callback_handle hotplugHandle;
+	UVQueue<HotPlug*> hotplugQueue;
+	Napi::ObjectReference hotplugThis;
+	std::map<libusb_device*, Device*> byPtr;
+	Napi::FunctionReference deviceConstructor;
+
+    ModuleData(): hotplugQueue(handleHotplug) {
+    }
```

Comment:
In order to clean up when our module is unloaded, we can either register a cleanup hook, but I think we can also add a destructor here that does the cleanup.

Cleanup tasks are (in order):

 1. Stop the hotplug polling thread. I'd do it by adding a `std::atomic<bool>` to `ModuleData`, and checking this boolean in the `while` condition from USBThreadFn. Then in the destructor set it to false, and join the thread. I think `detach` shouldn't be called if we do this.

 2. Destroy the libusb context by calling `libusb_exit` on `usb_context`. (only do this if usb_context != nullptr, and initialize usb_context to nullptr)

There's one detail I still have to think through, which is the garbage collection of the devices we created. It's possible we *do* need a cleanup hook in the end.

Reply:
Ok thanks, let me know your thoughts on the gc and I can look to implement

Reply:
The two tasks above will have to be implemented, that's for sure. But depending on how GC plays out, we may need to implement something more